tpm.go

```diff
@@ -0,0 +1,937 @@
+		if !xerrors.As(err, &unavailErr) && !xerrors.As(err, &verifyErr) {
+			return nil, xerrors.Errorf("cannot initialize TPM connection: %w", err)
+		}
+	}
+
+	succeeded = true
+	return t, nil
+}
+
+// SecureConnectToDefaultTPM will attempt to connect to the default TPM, verify the manufacturer issued endorsement key certificate
+// against the built-in CA roots and then verify that the TPM is the one for which the endorsement certificate was issued.
+//
+// If provided, the ekCertDataReader argument should read from a file or buffer created previously by FetchAndSaveEkCertificateChain,
+// SaveEkCertificateChain or EncodeEkCertificateChain. This makes it possible to connect to the TPM without requiring network access
+// to obtain certificates.
+//
+// If the data read from ekCertDataReader cannot be unmarshalled or parsed correctly, a EkCertVerificationError error will be
+// returned.
+//
+// If ekCertDataReader does not contain an endorsement key certificate, this function will attempt to obtain the certificate for the
+// TPM. This does not require network access. If this fails, a EkCertVerificationError error will be returned.
+//
+// If ekCertDataReader is nil, this function will attempt to obtain the endorsement key certificate for the TPM and then download the
+// parent certificates, which requires network access. If this fails, a EkCertVerificationError error will be returned. If network
+// access is unavailable and there is no existing EK certificate blob created previously by FetchAndSaveEkCertificateChain,
+// SaveEkCertificateChain or EncodeEkCertificateChain, then ConnectToDefaultTPM must be used instead.
+//
+// If verification of the endorsement key certificate fails, a EkCertVerificationError error will be returned. If the endorsement
+// key certificate data was supplied by the caller, this might mean that the data is invalid and should be recreated.
+//
+// In order for the TPM to prove it is the device for which the endorsement key certificate was issued, an endorsement key is
+// required. If the TPM doesn't contain a valid persistent endorsement key at the expected location (eg, if ProvisionTPM hasn't been
+// executed yet), this function will attempt to create a transient endorsement key. This requires knowledge of the endorsement
+// hierarchy authorization value, which will be empty on a newly cleared device. If there is no object at the persistent endorsement
+// key index and creation of a transient endorement key fails, ErrProvisioning will be returned.
+//
+// If the TPM cannot prove it is the device for which the endorsement key certificate was issued, a TPMVerificationError error will be
+// returned. This can happen if there is an object at the persistent endorsement key index but it is not the object for which the
+// endorsement key certificate was issued, and creation of a transient endorsement key fails because the correct endorsement hierarchy
+// authorization value hasn't been provided.
```

Comment:
all this doc doesn't mention endorsementAuth []byte which is a bit strange, also given it's []byte is not super obvious what exactly goes in there